### PR TITLE
Filter mutation frequencies tables

### PR DESCRIPTION
**Pull request by ewafula**
#### What scientific question is your analysis addressing?
Remove Ensembl (ESNG) gene identifier in the mutation frequency tables, including SNV, CNV and fusion that are not in GENCODE v38 and Ensembl package 104.

#### What was your approach?
- Filter out ENSGs in the mutation frequency tables, including SNV, CNV, and fusion that are not in GENCODE v38 and Ensembl package 104, and write out new TSV and JSONL mutation frequencies files.
- List filtered our ENSGs in the Rmd notebook for each mutation frequencies tables

#### What GitHub issue does your pull request address?
- https://github.com/PediatricOpenTargets/ticket-tracker/issues/268
- https://github.com/PediatricOpenTargets/ticket-tracker/issues/265
- https://github.com/PediatricOpenTargets/ticket-tracker/issues/264

#### Which areas should receive a particularly close look?
Review the entire code and test run the module. Also, check the results produced by the module  for consistency 

#### Is there anything that you want to discuss further?
No

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?
Yes.

#### What types of results are included (e.g., table, figure)?
TSV and JSONL tables

#### What is your summary of the results?
- `results/gene-level-snv-consensus-annotated-mut-freq.tsv.gz`
- `results/gene-level-snv-consensus-annotated-mut-freq.jsonl.gz`
- `results/variant-level-snv-consensus-annotated-mut-freq.tsv.gz`
- `results/variant-level-snv-consensus-annotated-mut-freq.jsonl.gz`
- `results/gene-level-cnv-consensus-annotated-mut-freq.tsv.gz`
- `results/gene-level-cnv-consensus-annotated-mut-freq.jsonl.gz`
- `results/putative-oncogene-fused-gene-freq.tsv.gz`
- `results/putative-oncogene-fused-gene-freq.jsonl.gz`
- `results/putative-oncogene-fusion-freq.tsv.gz`
- `results/putative-oncogene-fusion-freq.jsonl.gz`
- `filter-mutation-frequencies-tables.nb.html`

#### Reproducibility Checklist

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [ ] This analysis has been added to continuous integration.

#### Documentation Checklist

- [x] This analysis module has a `README` and it is up to date.
- [ ] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

### CODE REVIEW

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (runjin326) [inline analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]: A couple things here:
1) We would add the gencode v38 to data release, so could you please modify the path to data directory (also move your input there so that this module runs)? 
2) The one that we use is `primary assembly` to be consistent with our previous releases, so could you please also use `gencode.v38.primary_assembly.annotation.gtf.gz`? - It can be downloaded by running 
```
curl -O ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_38/gencode.v38.primary_assembly.annotation.gtf.gz"
```
 or download directly from the website. 

3) In CNV module, I actually use the gtf file by loading as following (which might be slightly faster) - but feel free to keep what you have here as well:
```
   gtf <- file.path(data_fir, "gencode.v38.primary_assembly.annotation.gtf.gz")
   gencode_gtf <- rtracklayer::import(con = gtf) %>% as.data.frame()
```

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (runjin326) [inline analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]: Looks like when I am running in the docker, the `compress = "auto"` errored as being an un-used argument. Could you please double check to see what is going on there?

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (runjin326) [inline analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]: Looks like the checks for what gets filtered out are not working for all tables. From the code it looks like you are asking which ENSG IDs are not in v38 AFTER the filtering step , which should be none if the filtering step works
Instead. I am interested in seeing which ones get filtered out so we should ask which ENSG IDs are not in v38 BEFORE filtering?
Additionally, could you please also output the `Kids_First_Biospecimen_ID` associated with those filtered ENSG IDs so that if necessary, we would have a record of what we might need to re-align manually? 
Thanks!

**Review by runjin326 — CHANGES_REQUESTED** [review]
@ewafula, thanks so much for preparing this! Overall, the code works and is doing what we intend to do. Just a few changes to get more QC data.

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (ewafula) [reply analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]: @runjin326, now using `gencode v38` file, `gencode.v38.primary_assembly.annotation.gtf.gz` from the repo data directory. I also updated the gtf loading module from `data.table::fread`  to  `rtracklayer::import`. But `data.table::fread` is faster even though rtracklayer::import` automatically adds column names to the data frame rather than adding them manually as I had done before. Thanks!

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (ewafula) [reply analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]: @runjin326, the package `data.table` in the OpenPedCan Docker is much older and does not support file compression as the current versions do.  I have replaced it with `reader::write_tsv`.

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (ewafula) [reply analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]: @runjin326, please carefully look at the filtering function, `filter_freq_table()` in `filter-mutation-frequencies-tables.Rmd`. It returns an unfiltered data frame of the frequencies tables. All filtering is done in the intermediate sections of the function code and written to file without altering the load data frame. Others there would have been no results of filtered ENSGs. 

How do we output `Kids_First_Biospecimen_ID` associated with those filtered `ENSGs`? The frequencies tables don't have a list of `sample IDs` associated with mutation `ENSGs`. Instead, I have included the associated gene symbols.

Comment (runjin326) [reply analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]: @ewafula, I see - sorry about the confusion!
So what I was thinking was - now that you have ENSGs, can you go back to the original table (i.e., `fusion-putative-oncogenic.tsv` for fusions, `consensus_wgs_plus_cnvkit_wxs.tsv.gz` for CNV and `snv-consensus-plus-hotspots.maf.tsv.gz` for SNV), find out those ENSGs and output the `Kids_First_Biospecimen_ID`s associated with those ENSGs - basically trying to figure out how many samples are affected and who are they.
Additionally, now that I am thinking about this, maybe also use `histologies.tsv`, filter to those samples, select cohorts and do a `table` call to get a summary of what cohorts they belong to.

Comment (ewafula) [reply analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]: Sure, @runjin326. Will do.

Comment (ewafula) [reply analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]: > Sure, @runjin326. Will do.

@runjin326, wrote dropped ENSG IDs to output files in addition to printing the tables out in the notebook. Too many associated Kids_First_Biospecimen_IDs for one to patiently page through.

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (runjin326) [inline analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd:139]: For all tables - could you please print out `filtered_ensg` here so that we can see the ENSG IDs being filtered?

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (runjin326) [inline analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]: For this output - could you please do:
```
fitlered_ensg %>% select(Gene_symbol, Ensembl_ID, cohort) %>% table()
```
Just to see what are the genes that are affected and we can keep all the individual samples output to tsv files as you have done here.

**Review by runjin326 — COMMENTED** [review]
@ewafula , thanks for making the modifications - everything looks as expected except for a few minor changes that I mentioned (basically just optimize the html notebook printout for better visualization).
After those changes are in, this should be ready for merge.

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (ewafula) [reply analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]: > For this output - could you please do:
> 
> ```
> fitlered_ensg %>% select(Gene_symbol, Ensembl_ID, cohort) %>% table()
> ```
> 
> Just to see what are the genes that are affected and we can keep all the individual samples output to tsv files as you have done here.

@runjin326, sorry that am understanding what you are suggesting. All the tables are printed in an RMD file. Please open this attached.  Is there any need to use `fitlered_ensg %>% select(Gene_symbol, Ensembl_ID, cohort) %>% table()` instead of simply `fitlered_ensg`. These three columns are the only columns in the tables. Any advantage to select them rather than just display the whole tables as rendered in this attached HTML file. 
[filter-mutation-frequencies-tables.nb.html.zip](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/files/7889758/filter-mutation-frequencies-tables.nb.html.zip)

Comment (ewafula) [reply analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]: > 

@runjin326,  I get your point now.  You want to exclude `Kids_First_Biospecimen_IDs` and only print them out to file because make the tables much larger.  Sorry for the misunderstanding! I will ASAP

Comment (ewafula) [reply analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]: > fitlered_ensg %>% select(Gene_symbol, Ensembl_ID, cohort) %>% table()

@runjin326, update done and PR updated.  Used this `filtered_ensg %>% select(Gene_symbol, Ensembl_ID, cohort) %>% distinct()` to remove duplicates when leaving out  `Kids_First_Biospecimen_ID` column.

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (runjin326) [inline analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd:76]: One additional thing: could you please also read in PMTL files and see whether the ENSGs that we filtered are in PMTL:
```
pmtl_df <- readr::read_tsv("../gene_match/input/PMTL_v1.1.tsv")
```

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (runjin326) [reply analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd:139]: And here, could you please do:
```
filtered_ensg

# check how many symbols are actually in the PMTL df
sum(unique(filtered_ensg$Ensembl_ID) %in% pmtl_df$Ensembl_ID)
# print those symbols when they are present 
filtered_ensg$Ensembl_ID %in% pmtl_df$Ensembl_ID

```

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (runjin326) [inline analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd:213]: I am actually thinking to get the number of samples in each cohort - so could you please modify to:
```
filtered_ensg %>% dplyr::select( Ensembl_ID, cohort) %>% 
  table()
# display the samples that have the ENSEMBL ID that are in PMTL list
filtered_ensg %>% dplyr::filter(Ensembl_ID %in% pmtl_df$Ensembl_ID) %>% 
select(Ensemble_ID, Kids_First_Biospecimen_ID) %>% distinct()
```

Comment (ewafula) [reply analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd:213]: > display the samples that have the ENSEMBL ID that are in PMTL list
> filtered_ensg %>% dplyr::filter(Ensembl_ID %in% pmtl_df$Ensembl_ID) %>% 
> select(Ensemble_ID, Kids_First_Biospecimen_ID) %>% distinct()

@runjin326, PR updated. For all the frequencies tables, removed ENSG IDs are not in the PMTL list.  `IGL-@, IGH@ IGH-@` genes (symbols) in the fusion tables are not displayed in the `ENSG-cohort` RMD rendered tables because they don't have associated `ENSG IDs`. But they are reported in the more comprehensive tables written to file, including `samples IDs`.

Comment (chinwallaa) [reply analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd:213]: for clarification - " removed ENSG IDs are not in the PMTL list" - or  is this removing ENSG IDs that are not in Ensemble pkg 104 and Gencode v38? or internal QC to identify any geneIDs that are in the PMTL list that would get removed due to the filter.

Comment (runjin326) [reply analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd:213]: I read the code - he mentioned `# Display samples in removed ENSEMBL IDs that are in PMTL list` and the code confirms it is an internal QC to identify any geneIDs that are in the PMTL list that would get removed due to the filter.

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (runjin326) [inline analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]:
```suggestion
filtered_ensg %>% dplyr::select(Ensembl_ID, cohort) %>% 
     table()
```
From my original comment, I do mean to just do a table call - which will give us a count for each cohort. Could you please commit the exact change I specified above?

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (runjin326) [inline analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]:
```suggestion
filtered_ensg %>% dplyr::select(Ensembl_ID, cohort) %>% 
  table()
```

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (runjin326) [inline analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]:
```suggestion
filtered_ensg %>% dplyr::select(Ensembl_ID, cohort) %>% 
  table()
```

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (runjin326) [inline analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]:
```suggestion
filtered_ensg %>% dplyr::select(Ensembl_ID, cohort) %>% 
  table()
```

File: analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd
Comment (runjin326) [inline analyses/filter-mutation-frequencies-tables/filter-mutation-frequencies-tables.Rmd]:
```suggestion
filtered_ensg %>% dplyr::select(Ensembl_ID, cohort) %>% 
  table()
```

**Review by runjin326 — COMMENTED** [review]
Hey @ewafula - thanks so much for making the changes. 
One last change is for the display, we do want to call `table()` since this will sum up the number like screenshot here (I screenshot half of the table):
image
Everything else look good - would be able to approve after this change.

**Comment by ewafula** [comment]
> Hey @ewafula - thanks so much for making the changes. One last change is for the display, we do want to call `table()` since this will sum up the number like screenshot here (I screenshot half of the table): image Everything else look good - would be able to approve after this change.

@runjin326, I have updated and summary tables are now rendered with frequencies counts as shown in your screenshot above.

**Comment by chinwallaa** [comment]
Thanks  @ewafula  - can you also point to the doc where the methods/tables are - that show the results of the intersect between GENCODE v38 and Emsemble package 104, and PMTLs.  I looked for these in the [README](https://github.com/ewafula/OpenPedCan-analysis/tree/filter-mutation-frequencies-tables) and the [MODULE-README](https://github.com/ewafula/OpenPedCan-analysis/tree/dev/analyses#modules-at-a-glance)

**Comment by ewafula** [comment]
> Thanks @ewafula - can you also point to the doc where the methods/tables are - that show the results of the intersect between GENCODE v38 and Emsemble package 104, and PMTLs. I looked for these in the [README](https://github.com/ewafula/OpenPedCan-analysis/tree/filter-mutation-frequencies-tables) and the [MODULE-README](https://github.com/ewafula/OpenPedCan-analysis/tree/dev/analyses#modules-at-a-glance)

@chinwallaa, GENECODE v38 includes IDs in Ensemble package 104. Therefore the intersect between them is all the annotated human gene models.  All PMTL Ensemble IDs are present in GENECODE v38 (Ensemble package 104) and none was filtered out as shown in the summary RMD notebook attached. Therefore, the intersect is all of them.  Is there a reason why you want all the intersects among these three (GENCODE v38, Emsemble package 104,  PMTLs) reported?  The PR has not been merged yet. Here is the PR README link and results are summarized in the RMD notebook (also attached below). 
- https://github.com/ewafula/OpenPedCan-analysis/tree/filter-mutation-frequencies-tables/analyses/filter-mutation-frequencies-tables

`base) 88665a17de0f:analyses wafulae$ head gencode.v38.primary_assembly.annotation.gtf`
```
##description: evidence-based annotation of the human genome (GRCh38), version 38 (Ensembl 104)
##provider: GENCODE
##contact: gencode-help@ebi.ac.uk
##format: gtf
##date: 2021-03-12
chr1    HAVANA  gene    11869   14409   .       +       .       gene_id "ENSG00000223972.5"; gene_type "transcribed_unprocessed_pseudogene"; gene_name "DDX11L1"; level 2; hgnc_id "HGNC:37102"; havana_gene "OTTHUMG00000000961.2";
chr1    HAVANA  transcript      11869   14409   .       +       .       gene_id "ENSG00000223972.5"; transcript_id "ENST00000456328.2"; gene_type "transcribed_unprocessed_pseudogene"; gene_name "DDX11L1"; transcript_type "processed_transcript"; transcript_name "DDX11L1-202"; level 2; transcript_support_level "1"; hgnc_id "HGNC:37102"; tag "basic"; havana_gene "OTTHUMG00000000961.2"; havana_transcript "OTTHUMT00000362751.1";
```
[filter-mutation-frequencies-tables.nb.html.zip](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/files/7979303/filter-mutation-frequencies-tables.nb.html.zip)

**Comment by chinwallaa** [comment]
>>GENECODE v38 includes IDs in Ensemble package 104
>> [README](https://github.com/ewafula/OpenPedCan-analysis/tree/filter-mutation-frequencies-tables/analyses/filter-mutation-frequencies-tables)
Thanks @ewafula

**Comment by ewafula** [comment]
@runjin326, this PR also synced with upstream dev and can merge if ready. I think @chinwallaa  already reviewed and needs to approve before merging.

**Comment by jharenza** [comment]
@ewafula did this code change result in new tables given to FNL for their current release of the MTP? if so, we will want to get this merged before the repo release to sync with the MTP. cc @chinwallaa @afarrel @taylordm @zdorman

**Comment by jharenza** [comment]
Note, this is not in the first release of the MTP, so holding off on merging until after repo release v1

**Pull request merged**
